Saule/Resources/IApiResourceProvider.cs

```diff
@@ -11,5 +11,13 @@ public interface IApiResourceProvider
         /// <param name="dataObject">Data object that is serialized. If it's an array, then it will be called for each item in the array</param>
         /// <returns>ApiResource that should be used for specific data object</returns>
         ApiResource Resolve(object dataObject);
+
+        /// <summary>
+        /// Returns ApiResource based on the data object and specific relationship
+        /// </summary>
+        /// <param name="dataObject">Data object that is serialized. If it's an array, then it will be called for each item in the array</param>
+        /// <param name="relationship">Relationship resource for which we are resolving api resource</param>
+        /// <returns>ApiResource that should be used for specific data object and relationship</returns>
+        ApiResource ResolveRelationship(object dataObject, ApiResource relationship);
```

Comment:
Can you explain how the relationship can affect the type? As far as I can tell, the type should be the same whether the resource is used from within a relationship or not.

The case where it is not seems like an unlikely scenario to me, but at the same time the cost to having this extra method is very low, so maybe it doesn't matter.

Reply:
yeah, it's our specific case that we want to achieve. the idea is to have base type and inherited types. all of them would be available on the same endpoint but type and attributes might vary depending on the type. The closest example would be unit tests that i've added where we have a `Group` and it has `List<Shape> Shapes` that are marked with `HasMany<ShapeResource>`. And each Shape instance might be either Circle or Rectangle

with my previous's PR i made it possible to have endpoint that returns `public IEnumerable<Shape> GetAll()` and then each Shape might be different inherited type and have additional attributes compared to base Shape and it works great. 

But if we have a different resource like Group that has a reference to Shapes, then without this change - it still would return base `shape` type and attributes of the base class. And original change wasn't enough to handle that case as we have `GroupResource` that has `HasMany<ShapeResource>(..)` and we didn't call that initial `Resolve(object dataObject)` method for that case so it always used `ShapeResource`.  And to handle it i've added second Resolve that also would have relationship object itself so when we resolve `ApiResource`  we can do that based on specific `ApiResource relationship` and it might be different


in case of our specific scenario, we have some kind of online menu that has groups and some menu items inside and there are some properties that are the same for all of them and they are on the base menu item type. But then depending on the item itself properties might vary so we want to return different inherited type and additional attributes based on type so it gives us more flexibility

Reply:
just want to add that in the initial PR that `Resolve` method wasn't called at all for `included`\ `relationships` nodes and it always used `ApiResource` from `HasMany`\ `BelongsTo` and there wasn't any way to override it without the change. 

And it wasn't enough just to call `Resolve(..)` method for that case as default implementation always return original `ApiResource` from `ReturnsResource` attribute and it won't be the same as in `apiResource.Relationships[x].RelatedResource`. That second method helps as we just pass `RelatedResource` itself as parameter and then default implementation returns it back. And the same time allow us to override it in our implementation as well

Reply:
Right, so this is needed for the default implementation, not the custom ones. Presumably they would use reflection or something instead.